---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
# knitr::opts_chunk$set(
#   collapse = TRUE,
#   comment = "#>",
#   out.width = "100%"
# )
```

# CSCpromoters

<!-- badges: start -->
<!-- badges: end -->

## About

The goal of CSCpromoters is to extract the sequences of promoter regions from the genome fasta files.


## Description

`CSCpromoters` allows users to extract the sequences of promoter regions from the genome fasta files. `CSCpromoters` can calculate the length of promoters based on the distance to the closest upstream gene. `CSCpromoters` also provides the option for users to enter the user-defined minimum and maximum promoter length.


## Installation

You can install the development version of `CSCpromoters` from 
[GitHub](https://github.com/) with:

```{r installation1, eval=FALSE}
# install.packages("devtools")
devtools::install_github("thiagomaf/CSCfasta")
devtools::install_github("thiagomaf/CSCpromoters")

# These commands will not work while the repositories are private.
# There are work-around, we can talk about it later.
```

Alternatively, you can download this git repository and install it locally. To 
do that, download the content of this repository by clicking the big green 
`Code` button (on the top right) and then click `Download ZIP`. Download it to 
a local folder. 

- [CSCpromoters](https://github.com/thiagomaf/CSCpromoters)

After that, run the following code line:

```{r installation2, eval=FALSE}
devtools::install_local("PATH/CSCpromoters-main.zip")
```


## Usage Example
This is a basic example which shows you how to solve a common problem:

### SETUP
#### Load libraries
```{r libraries}
library(magrittr)
library(CSCpromoters)
```

#### Define constants
```{r constants, eval=FALSE}
gff_folder <- paste0(
  "./"
)

fasta_list <- (function(
    .chr_list = c(
      "chr1H", "chr2H", "chr3H", "chr4H", "chr5H", "chr6H", "chr7H", "chrUn"
    ),
    .folder   = "./"
) {
  .folder %>% 
    paste0(paste0("Hordeum_vulgare.refseq[", .chr_list, "].fasta")) %>% 
    magrittr::set_names(.chr_list)
})()
```

```{r include = FALSE}
gff_folder <- paste0(
  "../../../",
  "Data/FASTA sequences/Genomes/Hordeum vulgare - Golden Promise/"
)

fasta_list <- (function(
    .chr_list = c(
      "chr1H", "chr2H", "chr3H", "chr4H", "chr5H", "chr6H", "chr7H", "chrUn"
    ),
    .folder   = paste0(
      "../../../",
      "rcs-gedo2-team_coldstorage/LAB_Share/RNAseq/barley/genome_GP/"
    )
) {
  .folder %>% 
    paste0(paste0("Hordeum_vulgare.refseq[", .chr_list, "].fasta")) %>% 
    magrittr::set_names(.chr_list)
})()
```


### LOAD DATA
#### H. vulgare cv. Golden Promise TxDB from GFF file
```{r gff}
txdb <- paste0(
  gff_folder,
  "Horvul_GP_v1r1_Apollo_30_06_20_named_product_GO.gff3"
) %>%
  make_txdb(
    .data_source = "Hv - Golden Promise",
    .organism    = "Hordeum vulgare"
  )

# For some reason this cannot be properly loaded from an .RData or .rda file,
# must be run on every new R session
```

#### H. vulgare cv. Golden Promise annotations
```{r annotations}
annotations <- txdb %>% get_txdb_annotation()
```


### GET PROMOTERS
#### Explicit pipeline
```{r explicit_pipeline}
annotations %>%
  # Choose which loci to use
  filter_locus(
    .keep = c(
      "chr6Hg0657981",
      "chr1Hg0040651",
      "chr3Hg0328761",
      "chr4Hg0417171",
      "chr3Hg0345511",
      "chr4Hg0383761",
      "chr5Hg0547031",
      "chr4Hg0380431",
      "chr4Hg0421931",
      "chr7Hg0679581"
    )
  ) %>% 
  # Calculate distances to closest upstream locus
  get_promoter_distances(.debug = F) %>%
  # Trim found upstream distances and define promoter lengths
  trim_promoter_distances(.min_size = 100, .max_size = 2000) %>%
  # Get promoter sequences
  get_promoter_sequences(.txdb = txdb, .FASTA_list = fasta_list)
```

#### Wrap-up function
```{r wrap_up}
annotations %>%
  # Wrap-up function
  get_promoters(
    .txdb       = txdb,
    .FASTA_list = fasta_list,
    .keep       = c(
      "chr6Hg0657981",
      "chr1Hg0040651",
      "chr3Hg0328761",
      "chr4Hg0417171",
      "chr3Hg0345511",
      "chr4Hg0383761",
      "chr5Hg0547031",
      "chr4Hg0380431",
      "chr4Hg0421931",
      "chr7Hg0679581"
    )
  )
```


## Dependencies
### Libraries
- magrittr
- tidyr
- plyr
- dplyr
- purrr
- progress
- GenomicFeatures
- CSCprimers
	
### Files
- `[gff_folder]`/Annotation_Golden_Promise_v1r1_Apollo_300620_mRNA.fasta
- `[gff_folder]`/Horvul_GP_v1r1_Apollo_30_06_20_named_product_GO.gff3
- `[folder2]`/Hordeum_vulgare.refseq[chr1H].fasta
- `[folder2]`/Hordeum_vulgare.refseq[chr2H].fasta
- `[folder2]`/Hordeum_vulgare.refseq[chr3H].fasta
- `[folder2]`/Hordeum_vulgare.refseq[chr4H].fasta
- `[folder2]`/Hordeum_vulgare.refseq[chr5H].fasta
- `[folder2]`/Hordeum_vulgare.refseq[chr6H].fasta
- `[folder2]`/Hordeum_vulgare.refseq[chr7H].fasta
- `[folder2]`/Hordeum_vulgare.refseq[chrUn].fasta
