---
title: "R Notebook"
output: none
---

# SETUP
```{r}
library(magrittr)
library(progress)
library(CSCprimers)
library(GenomicFeatures)
library(Rsamtools)
library(tidyr)
library(dplyr)

# There is an issue with the loading of libraries. Some conflict with dplyr (or tidyverse as a whole) and either the GenomicFeatures or Rsamtools
# Omit messages!???
```


# LOAD DATA
## H. vulgare cv. Golden Promise annotations
```{r}
annotations <- load_fasta(
  .filename = paste0(
    "../../../FASTA sequences/Genomes/Hordeum vulgare - Golden Promise/",
    "Annotation_Golden_Promise_v1r1_Apollo_300620_mRNA.fasta"
  )
) %>%
  get_fasta_annotation() %>% # Add progress bar
  mutate(begin  = as.numeric(begin)) %>% 
  mutate(end    = as.numeric(end)) %>% 
  mutate(len    = as.numeric(len)) %>% 
  mutate(strand = as.numeric(strand))
# consider moving this to inside the get_fasta_annotation() OR create a test 
# function to be used on the fly - this is assumed to be true for downstream 
# analyses.
```

## H. vulgare cv. Golden Promise TxDB from GFF file
```{r}
gffFile <- paste0(
  # "E:/OneDrive - University of Cambridge/CSC/",
  "C:/Users/ta507/OneDrive - University of Cambridge/CSC/",
  "Data/FASTA sequences/Genomes/Hordeum vulgare - Golden Promise/",
  "Horvul_GP_v1r1_Apollo_30_06_20_named_product_GO.gff3"
)

txdb <- makeTxDbFromGFF(
  file       = gffFile,
  dataSource = "Hv - Golden Promise",
  organism   = "Hordeum vulgare"
)
```

# GET PROMOTERS
```{r}
# annotations_modif <- annotations %>%
#   mutate(begin  = as.numeric(begin)) %>% 
#   mutate(end    = as.numeric(end)) %>% 
#   mutate(len    = as.numeric(len)) %>% 
#   mutate(strand = as.numeric(strand))


pb <- progress_bar$new(
  total = annotations %>% 
    select(locus_tag, chr) %>% 
    unique() %>% 
    nrow()
)

# This should become a function
# THIS MUST BE PARALLELIZED
#minyao_promoters <- annotations_modif %>%
minyao_promoters <- annotations %>%
  # subset(locus_tag %in% c("chr1Hg0000001","chr1Hg0000061", "chr1Hg0000221")) %>% # TO REMOVE
  group_by(locus_tag, chr) %>%
  (function(.data) {
    pb$tick(0)
    .data
  }) %>% 
  group_modify(.f = function(.each_locus, .keys) {
    # .results <- annotations_modif %>%
    .results <- annotations %>%
      pivot_longer(cols = c("begin", "end")) %>% # PIVOT TO LONGER - BEGIN AND END VALUES
      subset(chr == .keys$chr) %>%
      subset(locus_tag != .keys$locus_tag) %>%
      subset(case_when(
        # .each_locus$strand ==  1 ~ value < max(.each_locus$begin, .each_locus$end, na.rm = TRUE),
        # .each_locus$strand == -1 ~ value > min(.each_locus$begin, .each_locus$end, na.rm = TRUE),
        .each_locus$strand ==  1 ~ value < .each_locus$begin,
        .each_locus$strand == -1 ~ value > .each_locus$end,
        TRUE               ~ NA
      ))
    
    .output <- if(nrow(.results) != 0) {
      .results <- .results %>%
        subset(case_when(
          .each_locus$strand ==  1 ~ value == max(value, na.rm = TRUE),
          .each_locus$strand == -1 ~ value == min(value, na.rm = TRUE)
        )) %>%
        mutate(dist = case_when(
          .each_locus$strand ==  1 ~ .each_locus$begin - value,
          .each_locus$strand == -1 ~ value - .each_locus$end
        )) %>% 
        # mutate(
        #   dist = min(
        #     abs(value - c(.each_locus$begin, .each_locus$end)), na.rm = TRUE
        #   )
        # )
        select(locus_tag, dist)
      
      data.frame(
        closest_locus = .results$locus_tag, 
        dist          = .results$dist
      )
    } else {
      data.frame(
        closest_locus = NA_character_, 
        dist          = NA
      )
    }
    
    pb$tick()
    
    # .output %>% print()
    .output
  })

rm(pb)
```

```{r}
# This should become a function
minyao_promoters2 <- minyao_promoters %>% 
  ungroup() %>%
  (function(.data, .min_size = 100, .max_size = 2000) {
    .data %>%
      subset(dist >= .min_size) %>% 
      mutate(promoter_size = case_when(
        dist > .max_size ~ .max_size,
        TRUE ~ dist
      ))
  })
```

```{r}
# temp <- minyao_promoters %>%
#   ungroup() %>% 
#   select(locus_tag) %>% 
#   group_by(locus_tag) %>% 
#   count(locus_tag)
# 
# names <- temp %>% 
#       subset(n > 1) %>% 
#       pull(locus_tag)
# 
# minyao_promoters %>% 
#   subset(locus_tag %in% names)
# 
# annotations_modif %>% 
#   subset(locus_tag %in% c("chr1Hg0000811", "chr1Hg0000821"))
```


# GET PROMOTERS
```{r}
# pb2 <- progress_bar$new(
#   total = minyao_promoters2 %>%
#     dplyr::select(locus_tag, chr) %>% 
#     unique() %>% 
#     nrow()
# )

my_promoters <- minyao_promoters2 %>%
  dplyr::slice(1:100) %>% # Taking too long to do the whole table
  # (function(.data) {
  #   pb2$tick(0)
  #   .data
  # }) %>% 
  group_by(chr) %>% # 'chr' should probably be set on the fly
  group_map(.f = function(
    .data, .keys, .txdb = txdb,
    #.min_size = 100, .max_size = 2000, # not being used yet
    .debug = TRUE
  ) {
    # This is not working as expected, need to check whether the txdb exists/was assigned; tryCatch() maybe
    # if(!quote(.txdb) %>% as.character() %>% exists()) stop(".txdb is not set") 
    if(is.null(.txdb)) stop(".txdb is NULL")
    
    .each_chr <- .keys %>% pull(chr) %>% unique()
    # if(.debug) print(.each_chr) # became native to the function operation, below
    
    # Talk to me!
    print(.each_chr)
    
    # Trim promoter sizes
    #... to-do
    
    # Open FASTA file for each chromossome prior to extracting promoters in each
    .fasta_file <- paste0(
      # "E:/OneDrive - University of Cambridge/CSC/",
      "C:/Users/ta507/OneDrive - University of Cambridge/CSC/",
      "rcs-gedo2-team_coldstorage/LAB_Share/RNAseq/barley/genome_GP/",
      paste0("Hordeum_vulgare.refseq[", .each_chr, "].fasta")
    ) %>%
      Rsamtools::FaFile() %>% 
      open()
    
    # Set progress bar
    .pb <- progress_bar$new(
      total = .data %>%
        dplyr::select(locus_tag) %>% 
        unique() %>% 
        nrow()
    )
    
    # Initiate progress bar
    # Get promoters for each loci
    .promoters <- .data %>%
      (function(.data) {
        .pb$tick(0)
        .data
      }) %>% 
      group_by(locus_tag) %>% # 'locus_tag' must be assign on the fly
      group_map(.f = function(.each_data, .each_keys) {
        # 'locus_tag' and 'promoter_size' must be assign on the fly
        .each_loci       <- .each_keys %>% pull(locus_tag) %>% unique()
        .each_upstream   <- .each_data %>% pull(promoter_size) %>% unique()
        .each_downstream <- 0 # implement properly lazy ass! Either function argument or data column
        
        .each_promoters <- .txdb %>% 
          GenomicFeatures::transcripts(filter = list("tx_name" = .each_loci)) %>%
          GenomicFeatures::getPromoterSeq(
            subject    = .fasta_file,
            upstream   = .each_upstream,
            downstream = .each_downstream
          ) %>% 
          set_names(.each_loci)
        
        .pb$tick()
        
        .each_promoters
      })
    
    # Close FASTA file
    close(.fasta_file)
    
    # End progress bar
    # .pb$tick()
    # rm(.pb)
    
    .promoters %>% 
      do.call(what = c, arg = .)
  }) %>% 
  do.call(what = c, arg = .)
```

# EXPORT SEQUENCES
```{r}
output_file <- "data/promoters_Min-Yao.fasta"
writeXStringSet(my_promoters, filepath = output_file)

save(my_promoters, file = "data/my_promoter.rda")
```